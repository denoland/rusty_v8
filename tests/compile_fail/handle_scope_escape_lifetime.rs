// Copyright 2019-2020 the Deno authors. All rights reserved. MIT license.
use std::pin::pin;

pub fn main() {
  let mut isolate = v8::Isolate::new(mock());
  let scope1 = pin!(v8::HandleScope::new(&mut isolate));
  let mut scope1 = scope1.init();

  let _local = {
    let scope = pin!(v8::HandleScope::new(&mut scope1));
    let mut scope = scope.init();
    let scope3 = pin!(v8::HandleScope::new(&mut scope));
    let mut scope3 = scope3.init();
    let context = v8::Context::new(&mut scope3, v8::ContextOptions::default());
    let mut scope3 = v8::ContextScope::new(&mut scope3, context);
    let scope4 = pin!(v8::EscapableHandleScope::new(&mut *scope3));
    let mut scope4 = scope4.init();
    let value = v8::Integer::new(&mut scope4, 42);
    scope4.escape(value)
  };
}

fn mock<T>() -> T {
  unimplemented!()
}
