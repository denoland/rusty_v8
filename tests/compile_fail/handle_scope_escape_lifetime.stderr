error[E0597]: `scope2` does not live long enough
  --> tests/compile_fail/handle_scope_escape_lifetime.rs:12:44
   |
9  |   let _local = {
   |       ------ borrow later stored here
10 |     let scope2 = pin!(v8::HandleScope::new(&mut scope1));
11 |     let mut scope2 = scope2.init();
   |         ---------- binding `scope2` declared here
12 |     let scope3 = pin!(v8::HandleScope::new(&mut scope2));
   |                                            ^^^^^^^^^^^ borrowed value does not live long enough
...
20 |   };
   |   - `scope2` dropped here while still borrowed

error[E0597]: `scope3` does not live long enough
  --> tests/compile_fail/handle_scope_escape_lifetime.rs:15:44
   |
9  |   let _local = {
   |       ------ borrow later stored here
...
13 |     let mut scope3 = scope3.init();
   |         ---------- binding `scope3` declared here
14 |     let context = v8::Context::new(&mut scope3, v8::ContextOptions::default());
15 |     let mut scope3 = v8::ContextScope::new(&mut scope3, context);
   |                                            ^^^^^^^^^^^ borrowed value does not live long enough
...
20 |   };
   |   - `scope3` dropped here while still borrowed

error[E0716]: temporary value dropped while borrowed
  --> tests/compile_fail/handle_scope_escape_lifetime.rs:12:18
   |
9  |   let _local = {
   |       ------ borrow later stored here
...
12 |     let scope3 = pin!(v8::HandleScope::new(&mut scope2));
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
...
20 |   };
   |   - temporary value is freed at the end of this statement
   |
   = note: consider using a `let` binding to create a longer lived value
   = note: this error originates in the macro `pin` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0716]: temporary value dropped while borrowed
  --> tests/compile_fail/handle_scope_escape_lifetime.rs:10:18
   |
9  |   let _local = {
   |       ------ borrow later stored here
10 |     let scope2 = pin!(v8::HandleScope::new(&mut scope1));
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
...
20 |   };
   |   - temporary value is freed at the end of this statement
   |
   = note: consider using a `let` binding to create a longer lived value
   = note: this error originates in the macro `pin` (in Nightly builds, run with -Z macro-backtrace for more info)
